File: apps/hiveweave/lib/hiveweave/services/permission.ex
defmodule HiveWeave.Services.Permission do
  @moduledoc """
  Permission evaluation with 4-mode presets and glob pattern matching.
  
  Modes:
    - readonly: read-only tools (read_file, list_files, grep, read_skill, etc.)
    - readwrite: readonly + write/edit/bash tools
    - full: all tools allowed by default (ask for dangerous ones)
    - custom: user-defined allow/deny/ask lists
  
  Rule evaluation: deny → ask → allow (deny takes priority)
  Supports wildcard/glob matching for tool patterns.

  Parameter-level matching: patterns like `Bash(npm *)` match a tool name
  AND its serialized arguments (glob on the arg string). The tool-name part
  inside the parentheses is compared case-insensitively so `Bash(npm *)`
  matches the `bash` tool. Plain (paren-less) patterns keep the original
  case-sensitive exact/glob behavior.
  """

  require Logger

  @readonly_tools ~w(read_file list_files grep read_skill read_project_memory 
    read_charter read_goals get_project_time get_real_time question todowrite
    message_superior send_message write_work_log check_agent_status
    read_roster list_subordinates list_all_agents read_work_logs
    get_skill_detail list_available_skills list_available_mcp
    git_worktree_list git_worktree_status)

  @readwrite_tools @readonly_tools ++ ~w(write_file edit_file delete_file 
    move_file create_directory delete_directory search_files bash apply_patch
    write_memory report_completion dispatch_task hire_agent
    approve_work reject_work websearch run_code_review run_security_audit
    run_tests run_perf_audit run_full_review)

  @dangerous_tools ~w(bash apply_patch write_file edit_file delete_file 
    delete_directory git_worktree_create git_worktree_merge 
    git_worktree_remove dismiss_agent transfer_agent save_charter update_goals)

  # Shell tools with full command execution — denied for readonly mode even
  # with approval, since they grant unrestricted filesystem/process access.
  @shell_tools ~w(run_command bash)

  @doc """
  Evaluate whether a tool is allowed for an agent.
  Returns :allow, :deny, or :ask.

  `input` is the tool's input map. It is used for parameter-level pattern
  matching such as `Bash(npm *)`. It may be omitted for backward compatibility.
  """
  def evaluate(agent, tool_name, input \\ nil) do
    mode = get_permission_mode(agent)

    case mode do
      "full" -> check_rules(agent, tool_name, :allow, nil, input)
      "readwrite" -> check_rules(agent, tool_name, :allow, @readwrite_tools, input)
      "readonly" -> check_rules(agent, tool_name, :allow, @readonly_tools, input)
      "custom" -> evaluate_custom(agent, tool_name, input)
      _ -> :allow  # default: allow for unknown modes
    end
  end

  @doc """
  Evaluate using deny/ask/allow lists with glob matching.
  Order: deny → ask → allow (deny takes priority)

  `input` is the tool's input map, used for parameter-level patterns
  like `Bash(npm *)`.
  """
  def evaluate_custom(agent, tool_name, input \\ nil) do
    denied = get_denied_tools(agent)
    ask = get_ask_tools(agent)
    allowed = get_allowed_tools(agent)

    cond do
      matches_pattern?(tool_name, denied, input) -> :deny
      matches_pattern?(tool_name, ask, input) -> :ask
      matches_pattern?(tool_name, allowed, input) -> :allow
      true -> :ask  # default: ask if not in any list
    end
  end

  @doc """
  Glob pattern matching for tool names.
  Supports: * (any chars), ToolName(arg_pattern) (parameter-level match).

  `input` is the tool's input map. When a pattern contains parentheses, the
  arg_pattern is glob-matched against a serialized argument string derived
  from `input` (e.g. the bash `command`). Patterns without parentheses ignore
  `input` and behave as before (exact / glob match on the tool name only).
  """
  def matches_pattern?(tool_name, patterns, input \\ nil)

  def matches_pattern?(tool_name, patterns, input) when is_list(patterns) do
    args = extract_args_string(tool_name, input)
    Enum.any?(patterns, fn p -> single_match?(tool_name, p, args) end)
  end
  def matches_pattern?(_tool_name, _patterns, _input), do: false

  defp single_match?(tool_name, pattern, args) when is_binary(pattern) do
    cond do
      pattern == "*" -> true
      # Parameter-level pattern: "ToolName(arg_pattern)"
      has_param_pattern?(pattern) ->
        match_param_pattern?(tool_name, pattern, args)
      pattern == tool_name -> true
      String.contains?(pattern, "*") ->
        regex = glob_to_regex(pattern)
        Regex.match?(~r/^#{regex}$/, tool_name)
      true -> false
    end
  end

  # A param pattern looks like `ToolName(...)`: contains '(' and ends with ')'.
  defp has_param_pattern?(pattern) do
    String.contains?(pattern, "(") and String.ends_with?(pattern, ")")
  end

  defp match_param_pattern?(tool_name, pattern, args) do
    # Strict parse: ToolName(arg_pattern) where ToolName is word chars.
    case Regex.run(~r/^(\w+)\((.+)\)$/, pattern) do
      [_, rule_tool, arg_pattern] ->
        # Case-insensitive tool-name comparison so `Bash(npm *)` matches `bash`.
        if String.downcase(rule_tool) == String.downcase(tool_name) do
          regex = glob_to_regex(arg_pattern)
          Regex.match?(~r/^#{regex}$/, args)
        else
          false
        end

      _ ->
        false
    end
  end

  # Convert a glob pattern (with `*`) into a regex source string anchored full-match.
  defp glob_to_regex(pattern) do
    pattern
    |> Regex.escape()
    |> String.replace("\\*", ".*")
  end

  # Build the argument string matched by parameter-level patterns.
  # For `bash`, the natural argument is the command. For other tools, join
  # the input values with spaces. nil input yields "".
  defp extract_args_string(_tool_name, nil), do: ""
  defp extract_args_string("bash", input) when is_map(input) do
    # Check both string and atom keys — tool_executor may pass either form
    input["command"] || input[:command] || ""
  end
  defp extract_args_string(_tool_name, input) when is_map(input) do
    input
    |> Enum.map(fn
      {_k, v} when is_binary(v) -> v
      {_k, v} -> to_string(v)
    end)
    |> Enum.join(" ")
  end
  defp extract_args_string(_tool_name, input), do: to_string(input)

  @doc """
  Get the permission mode for an agent.
  """
  def get_permission_mode(agent) when is_map(agent) do
    Map.get(agent, :permission_type) || "executor"
  end

  defp get_denied_tools(agent) do
    parse_list(Map.get(agent, :denied_tools))
  end

  defp get_ask_tools(agent) do
    parse_list(Map.get(agent, :ask_tools))
  end

  defp get_allowed_tools(agent) do
    parse_list(Map.get(agent, :allowed_tools))
  end

  defp parse_list(nil), do: []
  defp parse_list(""), do: []
  defp parse_list(json) when is_binary(json) do
    case Jason.decode(json) do
      {:ok, list} when is_list(list) -> list
      _ -> []
    end
  rescue
    _ -> []
  end

  defp check_rules(agent, tool_name, default, preset, input) do
    denied = get_denied_tools(agent)
    ask = get_ask_tools(agent)
    allowed = get_allowed_tools(agent)

    cond do
      matches_pattern?(tool_name, denied, input) -> :deny
      # Shell tools are denied for readonly mode even if not in the denied list
      preset == @readonly_tools and tool_name in @shell_tools -> :deny
      matches_pattern?(tool_name, ask, input) -> :ask
      matches_pattern?(tool_name, allowed, input) -> :allow
      preset && tool_name in preset -> default
      preset && tool_name not in preset -> :ask
      true -> default
    end
  end
end

File: apps/hiveweave-py/src/hiveweave/services/permission.py
"""Permission service — tool permission evaluation with capability hard gates.

契约 08 + P0 Hard Gates:
- Evaluation order: hard capability deny → user deny → ask → allow → role/mode fallback
- allowed_tools MUST NOT elevate past hard capability denies
- 4 模式: readonly, readwrite, full, custom (executor fallback)
"""

import json
import re
from typing import Any

import structlog

from hiveweave.db import meta as meta_db
from hiveweave.services.policy import (
    infer_role_family,
    policy_service,
    tool_hard_deny,
)

logger = structlog.get_logger()

# ──────────────────────────────────────────────────────────────────────────
# Tool presets — visibility for LLM tool lists (not sole authorization)
# Hard capability in PolicyService is the real gate.
# ──────────────────────────────────────────────────────────────────────────

_BASE_TOOLS = frozenset({
    "grep", "question", "todowrite", "websearch", "webfetch",
    "calculate",
    "schedule_alarm", "list_alarms", "cancel_alarm",
    "review", "read_file", "list_files", "read_skill", "list_available_skills",
    "read_memory", "write_memory",
    "send_message", "message_superior", "message_subordinate",
    "message_peer", "message_team",
    "ask_agent", "notify_agent", "commit_turn", "defer_task_advance",
    "lookup_dev_server",
    "read_roster", "update_roster", "write_work_log",
    "request_review", "list_subordinates", "check_agent_status",
    "get_platform_state",
    "read_charter", "read_goals", "view_org_chart", "read_work_logs",
    "git_worktree_list", "git_worktree_status",
    "get_tasks",
    "claim_task", "update_task_status", "submit_task", "update_progress",
    "attest_doc_review",
    "look_at_image",
})

CEO_TOOLS = _BASE_TOOLS | frozenset({
    # CEO: 行政 + 里程碑验收 + 文档权（DOC_WRITE）。无写码/bash/test。
    # browse 可看产品。关闸走 waive_attestation(单条 taskId)；
    # 自己的出证仍不算 approve。不能一次关掉所有任务。
    "check_agent_progress",
    # write/edit 任意文档；源码/配置由 policy classify_write_kind 硬拒
    "write_file", "edit_file",
    "browse", "browse_main",
    "create_task", "dispatch_task", "review_task",
    "cancel_task", "unclaim_task", "reassign_task", "waive_attestation",
    "waive_merge",
    "save_charter", "update_goals", "dismiss_agent", "transfer_agent",
    # merge/remove 仅作升级兜底 —— create 仍由系统侧在 dispatch/hire 时建
    "git_worktree_merge", "git_worktree_remove",
    # 终验对用户说
    "message_user",
})

COORDINATOR_BUILDER_TOOLS = _BASE_TOOLS | frozenset({
    # 中层（设计者+接缝工）：协调权 + 写码权（写码收敛到叶子间接缝；
    # 与 executor 同契约拥有独立 worktree）。
    "write_file",
    "bind_skill", "unbind_skill",
    "create_task", "dispatch_task", "review_task",
    # 台账出口：废弃/释放误绑任务 + 豁免 attestation（policy 已映射 DISPATCH/REVIEW）
    "cancel_task", "unclaim_task", "reassign_task", "waive_attestation",
    "waive_merge",
    "save_charter", "update_goals", "dismiss_agent", "transfer_agent",
    # merge/remove only — create is system-side on dispatch/hire
    "git_worktree_merge", "git_worktree_remove",
    "start_dev_server",
    "stop_dev_server",
    # 写码/验证工具
    "edit_file", "apply_patch", "delete_file", "move_file",
    "create_directory", "delete_directory", "search_files",
    "bash", "bash_main", "run_command", "run_tests", "browse", "browse_main", "game_run_case", "game_run_case_main",
    "job_kill",
    "spawn_subagent",
    "generate_image",
    "git_worktree_checkpoint",
    "run_code_review", "run_security_audit", "run_perf_audit",
    "run_full_review",
    "request_code_audit",
    "assert_visual",
})

# Legacy alias — builder coordinator 即原 COORDINATOR_TOOLS 语义的超集。
COORDINATOR_TOOLS = COORDINATOR_BUILDER_TOOLS

HR_TOOLS = _BASE_TOOLS | frozenset({
    "hire_agent", "dismiss_agent", "transfer_agent",
    "list_agent_templates",
    "bind_skill", "unbind_skill",
    "write_file",
})

# Legacy name kept for imports/tests — executor listing after hard gates.
# Do NOT include hire/dispatch/bash elevation here for "readonly" meaning;
# PolicyService still hard-denies based on role family.
READONLY_TOOLS = _BASE_TOOLS | frozenset({
    "bash", "bash_main", "write_file", "browse", "browse_main", "assert_visual", "game_run_case", "game_run_case_main", "edit_file",
    "job_kill",
    "spawn_subagent",
    "generate_image",
    "bind_skill", "unbind_skill",
    "start_dev_server",
    "stop_dev_server",
    "run_tests", "apply_patch",
    "create_directory", "delete_file", "move_file", "search_files",
    "delete_directory",
    "run_code_review", "run_security_audit", "run_perf_audit", "run_full_review",
    "git_worktree_checkpoint",
})

READWRITE_TOOLS = READONLY_TOOLS | frozenset({
    "run_command",
    "request_code_audit",
})

ALL_TOOLS = READWRITE_TOOLS | COORDINATOR_BUILDER_TOOLS | HR_TOOLS | frozenset({
    "run_command",
    "git_worktree_create", "git_worktree_merge", "git_worktree_remove",
    "create_task", "dispatch_task", "review_task",
    "hire_agent", "dismiss_agent", "transfer_agent",
    "save_charter", "update_goals",
    "message_user",
})

# 工具本体（task_tools）无角色守卫；executor 滥用由两层拦截：
# policy 硬能力门（DISPATCH/REVIEW，executor 族不具备）+ 此集合 deny。
COORDINATOR_ONLY_TOOLS = frozenset({
    "create_task", "dispatch_task", "review_task",
    "cancel_task", "unclaim_task", "reassign_task", "waive_attestation",
    "waive_merge",
    "git_worktree_merge", "git_worktree_remove",
})

EXECUTOR_ONLY_TOOLS: frozenset[str] = frozenset()


class PermissionService:
    """Evaluates tool permissions with hard capability gates first."""

    async def get_permission_mode(self, agent_id: str) -> str:
        agent = await meta_db.get_agent_by_id(agent_id)
        if agent is None:
            return "custom"
        return agent.get("permission_mode") or "readonly"

    async def evaluate(
        self, agent_id: str, tool_name: str, tool_args: dict | None = None
    ) -> str:
        """Evaluate whether a tool is allowed. Returns 'allow'/'deny'/'ask'."""
        decision, _reason = await self.evaluate_detailed(
            agent_id, tool_name, tool_args
        )
        return decision

    async def evaluate_detailed(
        self, agent_id: str, tool_name: str, tool_args: dict | None = None
    ) -> tuple[str, str | None]:
        """Like evaluate, but also returns a deny reason for the model.

        Order: hard capability → user deny → ask → allow → role/mode fallback.
        ``allowed_tools`` cannot override hard capability deny.
        Returns ``(action, reason)`` — reason is set only on deny.
        """
        agent = await meta_db.get_agent_by_id(agent_id)
        if agent is None:
            return "ask", None

        hard = policy_service.hard_check(agent, tool_name, tool_args)
        if hard:
            logger.info(
                "policy.hard_deny",
                agent_id=agent_id,
                tool=tool_name,
                reason=hard[:200],
            )
            return "deny", hard

        mode = agent.get("permission_mode") or "readonly"
        denied = self._parse_list(agent.get("denied_tools"))
        ask = self._parse_list(agent.get("ask_tools"))
        allowed = self._parse_list(agent.get("allowed_tools"))

        if self._matches_pattern(tool_name, denied, tool_args):
            return (
                "deny",
                f"User denied_tools rule blocks '{tool_name}' "
                "(operator configured deny — not a role-capability gap)",
            )
        if self._matches_pattern(tool_name, ask, tool_args):
            return "ask", None
        if self._matches_pattern(tool_name, allowed, tool_args):
            return "allow", None

        family = infer_role_family(agent)
        permission_type = (agent.get("permission_type") or "").lower()

        if family == "hr":
            if tool_name in HR_TOOLS:
                return "allow", None
            return (
                "deny",
                f"Tool '{tool_name}' is not in the HR tool allowlist",
            )
        if family == "ceo":
            if tool_name in CEO_TOOLS:
                return "allow", None
            return (
                "deny",
                f"Tool '{tool_name}' is not in the CEO tool allowlist",
            )
        if family == "coordinator" or permission_type == "coordinator":
            if (
                tool_name in COORDINATOR_ONLY_TOOLS
                or tool_name in COORDINATOR_BUILDER_TOOLS
            ):
                return "allow", None
            return (
                "deny",
                f"Tool '{tool_name}' is not in the coordinator tool allowlist",
            )

        if tool_name in COORDINATOR_ONLY_TOOLS:
            return (
                "deny",
                f"Tool '{tool_name}' is coordinator-only "
                f"(role_family={family or 'unknown'})",
            )

        if mode == "full":
            return "allow", None
        if mode == "readwrite":
            if tool_name in READWRITE_TOOLS:
                return "allow", None
            return "ask", None
        if mode == "readonly":
            if tool_name in READONLY_TOOLS:
                return "allow", None
            return "ask", None
        if mode == "custom":
            return "ask", None
        return "ask", None

    def get_tools_for_mode(self, mode: str) -> list[str]:
        if mode == "readonly":
            return sorted(READONLY_TOOLS)
        if mode == "readwrite":
            return sorted(READWRITE_TOOLS)
        if mode == "full":
            return sorted(ALL_TOOLS)
        return []

    def get_tools_for_agent(self, agent: dict[str, Any]) -> list[str]:
        family = infer_role_family(agent)
        if family == "hr":
            names = HR_TOOLS
        elif family == "ceo":
            # 显式 ceo 分支 —— 禁止 unknown family 落 READWRITE 兜底泄漏工具。
            names = CEO_TOOLS
        elif family == "coordinator":
            names = COORDINATOR_BUILDER_TOOLS | COORDINATOR_ONLY_TOOLS
        elif family == "qa":
            names = READWRITE_TOOLS
        elif family == "executor":
            mode = agent.get("permission_mode") or "readwrite"
            names = set(self.get_tools_for_mode(
                mode if mode != "readonly" else "readwrite"
            ))
        else:
            # Unknown family — 最小暴露，不给 READWRITE 兜底。
            names = READONLY_TOOLS
        # 工具表可见性跟硬门对齐：列在表里但 capability 不够的不要发给模型（撞门）。
        visible = {n for n in names if tool_hard_deny(agent, n) is None}
        return sorted(visible)

    def _parse_list(self, raw: Any) -> list[str]:
        if not raw:
            return []
        try:
            result = json.loads(raw) if isinstance(raw, str) else raw
            return result if isinstance(result, list) else []
        except (json.JSONDecodeError, TypeError):
            return []

    def _matches_pattern(
        self, tool_name: str, patterns: list[str], tool_args: dict | None
    ) -> bool:
        if not patterns:
            return False
        args = self._extract_args_string(tool_name, tool_args)
        return any(self._single_match(tool_name, p, args) for p in patterns)

    def _single_match(self, tool_name: str, pattern: str, args: str) -> bool:
        if pattern == "*":
            return True
        if self._has_param_pattern(pattern):
            return self._match_param_pattern(tool_name, pattern, args)
        if pattern == tool_name:
            return True
        if "*" in pattern:
            regex = self._glob_to_regex(pattern)
            return re.match(f"^{regex}$", tool_name) is not None
        return False

    def _has_param_pattern(self, pattern: str) -> bool:
        return "(" in pattern and pattern.endswith(")")

    def _match_param_pattern(
        self, tool_name: str, pattern: str, args: str
    ) -> bool:
        m = re.match(r"^(\w+)\((.+)\)$", pattern)
        if not m:
            return False
        rule_tool, arg_pattern = m.group(1), m.group(2)
        if rule_tool.lower() != tool_name.lower():
            return False
        regex = self._glob_to_regex(arg_pattern)
        return re.match(f"^{regex}$", args) is not None

    def _glob_to_regex(self, pattern: str) -> str:
        return re.escape(pattern).replace(r"\*", ".*")

    def _extract_args_string(self, tool_name: str, tool_args: dict | None) -> str:
        if tool_args is None:
            return ""
        if tool_name in ("bash", "bash_main"):
            return str(tool_args.get("command", tool_args.get("cmd", "")))
        parts = []
        for v in tool_args.values():
            parts.append(v if isinstance(v, str) else str(v))
        return " ".join(parts)


permission_service = PermissionService()
